File: php/src/Types/Quote.php
<?php

declare(strict_types=1);

namespace K256\Types;

/**
 * Swap quote from K256.
 */
readonly class Quote
{
    public function __construct(
        /** Topic ID for the quote subscription */
        public string $topicId,
        /** Unix timestamp in milliseconds */
        public int $timestampMs,
        /** Sequence number */
        public int $sequence,
        /** Input token mint address */
        public string $inputMint,
        /** Output token mint address */
        public string $outputMint,
        /** Input amount in base units */
        public int $inAmount,
        /** Output amount in base units */
        public int $outAmount,
        /** Price impact in basis points */
        public int $priceImpactBps,
        /** Solana slot of the quote */
        public int $contextSlot,
        /** Algorithm used for routing */
        public string $algorithm,
        /** Whether this is an improvement over previous quote */
        public bool $isImprovement,
        /** Whether this quote is from cache */
        public bool $isCached,
        /** Whether this quote may be stale */
        public bool $isStale,
        /** JSON route plan */
        public ?string $routePlanJson,
    ) {}
}

File: rust/src/types/quote.rs
//! Quote types.

use serde::{Deserialize, Serialize};

/// Swap quote from K256.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct Quote {
    /// Input token mint address
    pub input_mint: String,
    /// Output token mint address
    pub output_mint: String,
    /// Input amount in base units
    pub in_amount: u64,
    /// Output amount in base units
    pub out_amount: u64,
    /// Price impact percentage
    pub price_impact_pct: f64,
    /// Solana slot of the quote
    pub slot: u64,
    /// Unix timestamp in milliseconds
    pub timestamp_ms: u64,
    /// List of route steps
    pub route_plan: Vec<serde_json::Value>,
    /// Minimum output (or max input for exactOut)
    #[serde(skip_serializing_if = "Option::is_none")]
    pub other_amount_threshold: Option<u64>,
    /// "ExactIn" or "ExactOut"
    #[serde(default = "default_swap_mode")]
    pub swap_mode: String,
}

fn default_swap_mode() -> String {
    "ExactIn".to_string()
}
